promql/engine.go

```diff
@@ -1325,16 +1380,101 @@ func (ev *evaluator) eval(expr parser.Expr) (parser.Value, storage.Warnings) {
 			newEv.interval = ev.noStepSubqueryIntervalFn(rangeMillis)
 		}
 
-		// Start with the first timestamp after (ev.startTimestamp - offset - range)
+		subqueryEvalTime := timestamp.FromFloatSeconds(e.Timestamp)
+		subqueryStartTime := ev.startTimestamp
+		if subqueryEvalTime != 0 {
+			if subqueryEvalTime > ev.endTimestamp {
+				newEv.endTimestamp = subqueryEvalTime - offsetMillis
+			}
+			if subqueryEvalTime < subqueryStartTime {
+				subqueryStartTime = subqueryEvalTime
+			}
+		}
+
+		// Start with the first timestamp after (subqueryStartTime - offset - range)
 		// that is aligned with the step (multiple of 'newEv.interval').
-		newEv.startTimestamp = newEv.interval * ((ev.startTimestamp - offsetMillis - rangeMillis) / newEv.interval)
-		if newEv.startTimestamp < (ev.startTimestamp - offsetMillis - rangeMillis) {
+		newEv.startTimestamp = newEv.interval * ((subqueryStartTime - offsetMillis - rangeMillis) / newEv.interval)
+		if newEv.startTimestamp < (subqueryStartTime - offsetMillis - rangeMillis) {
 			newEv.startTimestamp += newEv.interval
 		}
 
 		res, ws := newEv.eval(e.Expr)
 		ev.currentSamples = newEv.currentSamples
 		return res, ws
+	case *parser.ConstantExpr:
+		switch ce := e.Expr.(type) {
+		case *parser.StringLiteral, *parser.NumberLiteral:
+			return ev.eval(ce)
+		}
+
+		if e.Result != nil {
+			return e.Result, nil
+		}
+		newEv := &evaluator{
+			startTimestamp:           ev.startTimestamp,
+			endTimestamp:             ev.startTimestamp,
+			interval:                 ev.interval,
+			ctx:                      ev.ctx,
+			currentSamples:           ev.currentSamples,
+			maxSamples:               ev.maxSamples,
+			logger:                   ev.logger,
+			lookbackDelta:            ev.lookbackDelta,
+			noStepSubqueryIntervalFn: ev.noStepSubqueryIntervalFn,
+		}
+
+		canExtrapolate := canExtrapolateConstantExpr(e.Expr)
+		if !canExtrapolate {
+			newEv.endTimestamp = ev.endTimestamp
+		}
+
+		res, ws := newEv.eval(e.Expr)
+		ev.currentSamples = newEv.currentSamples
```

Comment:
What is res.val? And I don't see anything that is to be subtracted as we are only extrapolating below

Reply:
The result of the expression. What's extrapolation to do with anything?

Reply:
So we don't count the samples in the result?

Reply:
They'll already be counted within the eval, however when are they being uncounted?

Reply:
I am still confused about this samples limit. Is it the number of TSDB samples touched or held in the memory?

Reply:
And why do we need to subtract res.val if it is always held in the memory? (assuming the max samples in the samples in the memory)

Reply:
In memory, and once calculations involving the node are complete we no longer have it in memory - so it needs to be subtracted.

Reply:
Looking at how we have handled it for subquery, it's `ev.currentSamples = newEv.currentSamples`, so I thought it would work out here too since it comes subtracted? I will verify it (in that case sample calculation for subquery might also be wrong?)

Reply:
Since we are caching the result here, should we not avoid subtracting?

Reply:
The unittests should cover this one way or the other.